GHA: Add OCaml trunk to the test matrix

kit-ty-kate opened the pull request:
See #6670 
~Depends on https://github.com/ocaml/opam-repository/pull/28455~

The job is allowed to fail as bugs and uncaught unwanted behaviours are common at this point

kit-ty-kate committed:
GHA: Ensure curl's exit status to be non-zero on failure

rjbou commented:
Discussed on dev meeting: the PR is almost ready, there is an issue with ignoring error on OCaml trunk jobs with Cygwin.

kit-ty-kate committed:
GHA: Add OCaml trunk to the CI matrix

The job is allowed to fail as bugs and uncaught unwanted behaviours are common at this point
